### PR TITLE
GH-36886: [C++] Configure `azurite` in preparation for testing Azure C++ filesystem

### LINKED ISSUE #36886: [C++] Configure `azurite` in preparation for testing Azure C++ filesystem
### Describe the enhancement requested

We need to write tests for https://github.com/apache/arrow/issues/18014. `azurite` is like a fake Azure blob storage so it can be used to write integration tests. I think this should unblock https://github.com/apache/arrow/issues/29847

The work was basically all done by https://github.com/apache/arrow/pull/12914 but we need to extract the `azurite` stuff out of that PR so we can merge it incrementally in small PRs. 

**Related Issues:**
- https://github.com/apache/arrow/issues/18014 (is a child of)
- https://github.com/apache/arrow/issues/29847 (is depended upon by)

### Component(s)

C++

### PR DESCRIPTION
**Pull request by Tom-Newton**
### Rationale for this change
We need to write tests for https://github.com/apache/arrow/issues/18014. azurite is like a fake Azure blob storage so it can be used to write integration tests

### What changes are included in this PR?
Extract the `azurite` related changes from https://github.com/apache/arrow/pull/12914 to create a smaller PR that's easier to review. I have made very minimal changes compared to that PR. 

Currently `azurite` is configured for all the environments where `ARROW_AZURE` was enabled by https://github.com/apache/arrow/pull/35701. I assume its deliberate that its not enabled yet for windows, alpine, conda, debian or fedora builds. 

### Are these changes tested?
Its tested by there aren't really any good tests in this PR. I used this `azurite` config in https://github.com/apache/arrow/pull/36835 to make an integration test that uses the Azure C++ SDK. On its own we can't really write tests for this `azurite` setup PR. 

### Are there any user-facing changes?
No

* Closes: #36886

### CODE REVIEW

File: cpp/src/arrow/filesystem/azurefs_test.cc [hunk 1]
```diff
@@ -16,6 +16,8 @@
 // under the License.
 
 #include "arrow/filesystem/azurefs.h"
+#include <boost/process.hpp>
```
Comment (kou) [inline cpp/src/arrow/filesystem/azurefs_test.cc:19]: Could you use https://github.com/apache/arrow/blob/7cbbd3ee95bc0f7d20bb358e2978e2eb18a05304/cpp/src/arrow/filesystem/gcsfs_test.cc#L18-L35 ?

File: cpp/src/arrow/filesystem/azurefs_test.cc [hunk 2]
```diff
@@ -27,15 +29,76 @@
...
+    account_key_ =
+        "Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/"
+        "KBHBeksoGMGw==";
+    auto exe_path = bp::search_path("azurite");
+    if (exe_path.empty()) {
+      auto error = std::string("Could not find Azurite emulator.");
+      status_ = Status::Invalid(error);
+      return;
+    }
+    auto temp_dir_ = TemporaryDir::Make("azurefs-test-").ValueOrDie();
```
Comment (kou) [inline cpp/src/arrow/filesystem/azurefs_test.cc:54]:
```suggestion
    auto temp_dir_ = *TemporaryDir::Make("azurefs-test-");
```

File: cpp/src/arrow/filesystem/azurefs_test.cc (hunk 1 above)
Comment (Tom-Newton) [reply cpp/src/arrow/filesystem/azurefs_test.cc:34]: I don't see a reason not to

**Review by kou — APPROVED** [review]
+1

**Pull request merged**